src/main/scala/org/constellation/gossip/GossipService.scala

```diff
@@ -1,47 +1,123 @@
 package org.constellation.gossip
 
-import cats.data.OptionT
-import cats.effect.Concurrent
-import cats.implicits._
+import cats.Parallel
+import cats.effect.{Concurrent, Timer}
+import cats.syntax.all._
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import org.constellation.gossip.bisect.bisectA
 import org.constellation.gossip.sampling.PeerSampling
-import org.constellation.infrastructure.p2p.ClientInterpreter
+import org.constellation.gossip.state.{GossipMessage, GossipMessagePathTracker}
 import org.constellation.infrastructure.p2p.PeerResponse.PeerClientMetadata
 import org.constellation.p2p.Cluster
-import org.constellation.schema.{Id, NodeState}
+import org.constellation.schema.Id
 
-abstract class GossipService[F[_], A](selfId: Id, peerSampling: PeerSampling[F, Id], cluster: Cluster[F])(
-  implicit val F: Concurrent[F]
+import scala.concurrent.duration._
+
+abstract class GossipService[F[_]: Parallel, A](
+  selfId: Id,
+  peerSampling: PeerSampling[F],
+  cluster: Cluster[F],
+  messageTracker: GossipMessagePathTracker[F, A]
+)(
+  implicit F: Concurrent[F],
+  T: Timer[F]
 ) {
+  protected val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+
+  /**
+    * How to spread the message to peer
+    */
+  protected def spreadFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Unit]
+
+  /**
+    * How to check if message was received by peer
+    */
+  protected def validationFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Boolean]
+
+  def recover(message: GossipMessage[A]): F[Unit] = {
+    val sequence = message.path.toIndexedSeq
+
+    for {
+      failureNode <- bisectA(
+        (id: Id) => getClientMetadata(id).flatMap(validationFn(_, message))
+      )(sequence)
+      sequenceToRetry = sequence.dropWhile(failureNode.contains)
+      clients <- sequenceToRetry.toList.traverse(getClientMetadata)
+      _ <- clients.traverse(spreadFn(_, message))
+    } yield ()
+  }
 
-  protected def spread(data: A, fanout: Int)(spreadFn: (PeerClientMetadata, GossipMessage[A]) => F[Unit]): F[Unit] =
+  /**
+    * Start spreading the message to fanout
+    */
+  def spread(data: A, fanout: Int): F[Unit] =
     for {
+      _ <- logger.debug(s"Starting spreading of data: $data with fanout $fanout")
       messages <- peerSampling
         .selectPaths(fanout)
-        .map(_.map(GossipPath(_)).map(GossipMessage(data, _)))
+        .map(_.map(GossipMessage(data, _)))
+      _ <- messages.map { msg =>
```

Comment:
We don't need this. The proposer peer only sends proposals to the first layer in the partition map. Each peer, when it receives a proposal, just sends to the next peer in the partition map.

Reply:
It works like that.`messages` contains first layer only.

For example if you have following partitions and you are node `A` (initial proposer)

A -> B -> C
A -> D -> E
A -> F

Then messages contains following messages: `GossipMessage(A -> B), GossipMessage(A -> D), GossipMessage(A -> F)`

And we just "fire and forget" these messages in parallel. Each `GossipMessage` contains whole path, current position on path so nodes `B`, `D, `F` just spread to next nodes and move the current position on path.

Reply:
That being said for A -> B -> C the whole flow looks like this:



Node A:
Send `GossipMessage(path: A -> B -> C, curr: A, next: B)` to B

Node B:
Receive `GossipMessage(path: A -> B -> C, curr: A, next: B)`, update path by moving cursors and send `GossipMessage(path: A -> B -> C, curr: B, next: C)` to C

Node C:
Receive `GossipMessage(path: A -> B -> C, curr: B, next: C)`, update path by moving cursors and send `GossipMessage(path: A -> B -> C, curr: C, next: A)` to C

Node A:
Finish the cycle and confirm that path succeeded.

Reply:
Let me know if this is what you meant or not :)